src/std/io/bio/input.ss

```diff
@@ -105,6 +105,130 @@
+        (put-back! buf new-rlo previous-input)
+        (set! (&input-buffer-rlo bio) new-rlo)
+        (void)))
+     ((fx> rlo 0)
+      ;; we need to move the buffer contents to the right
+      (let* ((shift (fx- prevlen rlo))
+             (rhi+shift (fx+ rhi shift))
+             (buflen (u8vector-length buf)))
+        (if (fx> rhi+shift buflen)
+          ;; uh oh we need to grow the buffer; do it by a page
+          (let (new-buflen (fx+ buflen 4096))
+            (while (fx< new-buflen rhi+shift)
+              ;; ok not enough, add more pages (very unlikely, but still)
+              (set! new-buflen (fx+ new-buflen 4096)))
+            (let (new-buf (make-u8vector new-buflen))
+              (subu8vector-move! buf rlo rhi new-buf prevlen)
+              (put-back! new-buf 0 previous-input)
+              (set! (&input-buffer-buf bio) new-buf)
+              (set! (&input-buffer-rhi bio) rhi+shift)
+              (void)))
+          (begin
+            (subu8vector-move! buf rlo rhi buf prevlen)
+            (put-back! buf 0 previous-input)
+            (set! (&input-buffer-rlo bio) 0)
+            (set! (&input-buffer-rhi bio) rhi+shift)
+            (void)))))
+     ;; rlo=0
+     ((fx> rhi 0)
+      ;; we need to move the buffer contents to the right
+      (let ((rhi+shift (fx+ rhi prevlen))
+            (buflen (u8vector-length buf)))
+        (if (fx> rhi+shift buflen)
+          ;; uh oh we need to grow the buffer; do it by a page
+          (let (new-buflen (fx+ buflen 4096))
+            (while (fx< new-buflen rhi+shift)
+              ;; ok not enough, add more pages (very unlikely, but still)
+              (set! new-buflen (fx+ new-buflen 4096)))
+            (let (new-buf (make-u8vector new-buflen))
+              (subu8vector-move! buf 0 rhi new-buf prevlen)
+              (put-back! new-buf 0 previous-input)
+              (set! (&input-buffer-buf bio) new-buf)
+              (set! (&input-buffer-rhi bio) rhi+shift)
+              (void)))
+          (begin
+            (subu8vector-move! buf 0 rhi buf prevlen)
+            (put-back! buf 0 previous-input)
+            (set! (&input-buffer-rhi bio) rhi+shift)
+            (void)))))
+     ;; rlo=rhi=0
+     (else
+      (let (buflen (u8vector-length buf))
+        (if (fx> prevlen buflen)
+          ;; uh oh we need to grow the buffer; do it by a page
+          (let (new-buflen (fx+ buflen 4096))
+            (while (fx< new-buflen prevlen)
+              ;; ok not enough, add more pages (very unlikely, but still)
+              (set! new-buflen (fx+ new-buflen 4096)))
+            (let (new-buf (make-u8vector new-buflen))
+              (put-back! new-buf 0 previous-input)
+              (set! (&input-buffer-buf bio) new-buf)
+              (set! (&input-buffer-rhi bio) prevlen)
+              (void)))
+          (begin
+            (put-back! buf 0 previous-input)
+            (set! (&input-buffer-rhi bio) prevlen)
+            (void))))))))
```

Comment:
That makes consecutive put-backs expensive. Maybe worth noting in the documentation?

Reply:
yeah, thats why the list.
Maybe worth adding a note.

Reply:
nah, it is not really deterministic on how much it will cost; it is usually cheap, but sometimes it can be expensive.